### ISSUE TITLE
"cl" is required to be in PATH even for --backend vs2017

### ISSUE THREAD

**Issue by mojca** [mesonbuild/meson#2622]
After some successful usage of meson on "unix", I tried to use it on Windows. I have Visual Studio 2017 installed, but getting this to work is quite problematic.

The `vs2017` backend creates a project file, so it's not entirely clear to me why meson needs to have access to `cl` from `PATH`. And even if it does: couldn't it somehow find it itself given that it knows it should be looking for the binary inside VS2017 installation? In conrast `cmake` never asks for `cl` to be present when generating Visual Studio projects.

The most weird thing is that generating the project files for Visual Studio worked just fine on Mac. But I guess it didn't hardcode `clang` compiler into the project? An additional (independent) problem is that I didn't install all SDKs an the one that `meson` required was not available (it was an older Windows 8 SDK), so I had to manually retarget all projects. And for some reason I was unable to even run the build of `REGEN` afterwards due to `Access to the path 'z:\path\to\project\build\msvs\meson-private\' is denied.` (this was generated on Mac).

I'll try to figure out where `cl` lives and add it to path to see if that works any better.

**Comment by nirbheek** [comment]
The way to use Meson on Windows with MSVC currently is to run it from inside the Visual Studio toolchain console. That calls `vcvarsall.bat` for you with the correct arguments and sets the environment needed to use MSVC.

cmake does not require this because it handles all that for you. We should indeed add something similar to cmake's `-G` argument so you don't have to do this. Note, however, that `--backend=vs*` is required to generate vcxproj files, but you can also use the ninja backend with MSVC.

**Comment by gulgi** [comment]
Adding a copy of https://github.com/Microsoft/vswhere to the windows version would solve all that, right?

**Comment by nirbheek** [comment]
> Or maybe even better would be to have a generic --backend=vs that creates the correct visual studio project based on the environment.

fwiw, this already exists.

> Adding a copy of https://github.com/Microsoft/vswhere to the windows version would solve all that, right?

We can't use that code directly, but thanks for pointing it out, we should be able to look at it to figure out the best way to find Visual Studio.

**Comment by nirbheek** [comment]
Jonathan Blow published a C++ file (MIT licensed) that does the same job as vswhere with easier to understand code: 

https://pastebin.com/3YvWQa5c

See `find_visual_studio_by_fighting_through_microsoft_craziness()` in that paste. We may need to do some COM API calls through Python.

**Comment by nirbheek** [comment]
The target arch would be specified via the cross-file. The SDK and VS version could be in one option: `--vs=2017:universal`, or the SDK version could also be a property in the cross file.

**Comment by Sahnvour** [comment]
Is this still considered ? It would make working on windows much easier.
Requiring the use of VS' developer command prompts (vcvarsall.bat) is understandable but not very nice in day to day usage.

For example, one can't use nicer shells like git bash to issue build commands.

@nirbheek the vswhere project provides self-contained executables, is that not usable by Meson ? Alternatively, it's shipped in every Visual Studio installation at a fixed path, see https://github.com/microsoft/vswhere/wiki/Installing.
There's also a quite simple [implementation](https://github.com/ziglang/zig/blob/a5a53a182af3715edd9474fe94cad2223a6c7fb9/src/windows_sdk.cpp#L49) in the Zig compiler which I believe inspired Blow's version.

**Comment by nirbheek** [comment]
> Is this still considered ? It would make working on windows much easier.
> Requiring the use of VS' developer command prompts (vcvarsall.bat) is understandable but not very nice in day to day usage.

Yes, I would still like to see it. Particularly since cross-compilation is really difficult without it, and you definitely can't have both native and cross compilation with visual studio in the same project (for building a build tool that is then used to generate code to be built for cross-compilation) without this.

> @nirbheek the vswhere project provides self-contained executables, is that not usable by Meson ? Alternatively, it's shipped in every Visual Studio installation at a fixed path, see https://github.com/microsoft/vswhere/wiki/Installing.

`vswhere` is only available on the system if you install `Visual Studio 20(17|19) (Community|Enterprise|Professional|Preview)`. So it won't be available if you only have `VS 2015` or older installed, or if you have `VS 20(17|19) Build Tools` installed. In those cases you could look in the default installation paths (which are different on 32-bit, 64-bit, and ARM 64-bit). But we will need a way to override this, because people do install in strange places.

I've already implemented all this for gstreamer's cerbero meta-build-wrapper: https://gitlab.freedesktop.org/gstreamer/cerbero/-/blob/master/cerbero/ide/vs/env.py, so we can just copy that code. I'm explicitly giving permission to use it under Apache-2.0, same as Meson in case someone else wants to work on that. I will likely not get to it anytime soon 🙂

**Comment by naveen521kk** [comment]
I slightly edited it to get it working with `Visual Studio 2019 Build Tools`.
```patch
diff --git a/mesonbuild/mesonmain.py b/mesonbuild/mesonmain.py
index ce59ba438..b751e670c 100644
--- a/mesonbuild/mesonmain.py
+++ b/mesonbuild/mesonmain.py
@@ -62,21 +62,31 @@ def setup_vsenv():
     if shutil.which('cl.exe'):
         return

-    bat_locator_bin = r'c:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe'
-    if not os.path.exists(bat_locator_bin):
+    root = os.environ.get("ProgramFiles(x86)") or os.environ.get("ProgramFiles")
+    bat_locator_bin = pathlib.Path(root, 'Microsoft Visual Studio/Installer/vswhere.exe')
+    if not bat_locator_bin.exists():
         return
-    bat_json = subprocess.check_output([bat_locator_bin, '-latest', '-format', 'json'])
+    bat_json = subprocess.check_output(
+        [
+            str(bat_locator_bin),
+            '-latest',
+            '-prerelease',
+            '-requiresAny',
+            '-requires', 'Microsoft.VisualStudio.Component.VC.Tools.x86.x64',
+            '-requires', 'Microsoft.VisualStudio.Workload.WDExpress',
+            '-products', '*',
+            '-format',
+            'json'
+        ]
+    )
     bat_info = json.loads(bat_json)
-    if not bat_info:
-        bat_json = subprocess.check_output([bat_locator_bin, '-prerelease', '-format', 'json'])
-        bat_info = json.loads(bat_json)
     if not bat_info:
         # VS installer instelled but not VS itself maybe?
         return
     print('Activating VS', bat_info[0]['catalog']['productDisplayVersion'])
-    bat_root = bat_info[0]['installationPath']
-    bat_path = bat_root + r'\VC\Auxiliary\Build\vcvars64.bat'
-    if not os.path.exists(bat_path):
+    bat_root = pathlib.Path(bat_info[0]['installationPath'])
+    bat_path = bat_root / 'VC/Auxiliary/Build/vcvars64.bat'
+    if not bat_path.exists():
         return

     bat_file = pathlib.Path.home() / 'vsdetect.bat'
```
Got the idea from [setuptools](https://github.com/pypa/setuptools/blob/a5131f0b82e098da6c07a03a47f36f3a52f73fb6/setuptools/msvc.py#L176). Should I make a PR?

**Comment by naveen521kk** [comment]
> @naveen521kk makes sense at first look, I'm not sure we require Microsoft.VisualStudio.Workload.WDExpress but that can be discussed in a PR.

Ok made https://github.com/mesonbuild/meson/pull/8808
